### ISSUE TITLE
HighDPI 200% scaling setting not persisted between sessions.

### ISSUE THREAD

**Issue by SneWs** [pop-os/pop#320]
NAME="Pop!_OS"
VERSION="18.04 LTS"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Pop!_OS 18.04 LTS"
VERSION_ID="18.04"
HOME_URL="https://system76.com/pop"
SUPPORT_URL="http://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

**Issue/Bug Description:**
On my main machine I have a 1080Ti with 2 4k screens (ASUS) connected over HDMI 2.0. HighDpi is properly and automatically activated, but the scale factor is still 100% out of box, and once changed to 200% everything looks crisp and very much readable and fully functional.

However, once I logout and back in or reboot the machine the 200% scale factor ain't restored from settings and I need to re-apply the setting every time I log back into the machine. 

**Steps to reproduce (if you know):**
With the hardware described above, go into settings > devices > display and change the scaling factor to 200%. Log out or reboot and log back in and the setting has been reverted to 100%.

**Expected behavior:**
The scale factor should be persisted between sessions.

**Other Notes:**
This is my big main rig thus I won't experiment too much with new kernels or drivers, I can however try to help repro and produce log files the best I can.

**Comment by djordan2** [comment]
Please run `xrandr` and provide the output.  That should help see why the system is setting 100% scaling.  Many 4k desktop monitors aren't actually dense enough to be automatically classified as hidpi, so that may be part of it.

**Comment by SneWs** [comment]
```
xrandr
Screen 0: minimum 8 x 8, current 7680 x 2160, maximum 32767 x 32767
DVI-D-0 disconnected (normal left inverted right x axis y axis)
HDMI-0 connected 3840x2160+3840+0 (normal left inverted right x axis y 
axis) 596mm x 335mm
    3840x2160     60.00*+  59.94    50.00    30.00    29.97 25.00    23.98
    2560x1600     59.97
    2560x1440     59.95
    1920x1080     60.00    59.94    50.00    29.97    25.00 23.98    
60.00    50.04
    1680x1050     59.95
    1440x900      59.89
    1440x576      50.00
    1440x480      59.94
    1280x1024     75.02    60.02
    1280x960      60.00
    1280x720      60.00    59.94    50.00
    1152x864      75.00
    1024x768      75.03    70.07    60.00
    800x600       75.00    72.19    60.32    56.25
    720x576       50.00
    720x480       59.94
    640x480       75.00    72.81    59.94    59.93
    640x350       70.10
HDMI-1 connected primary 3840x2160+0+0 (normal left inverted right x 
axis y axis) 596mm x 335mm
    3840x2160     60.00*+  59.94    50.00    30.00    29.97 25.00    23.98
    2560x1600     59.97
    2560x1440     59.95
    1920x1080     60.00    59.94    50.00    29.97    25.00 23.98    
60.00    50.04
    1680x1050     59.95
    1440x900      59.89
    1440x576      50.00
    1440x480      59.94
    1280x1024     75.02    60.02
    1280x960      60.00
    1280x720      60.00    59.94    50.00
    1152x864      75.00
    1024x768      75.03    70.07    60.00
    800x600       75.00    72.19    60.32    56.25
    720x576       50.00
    720x480       59.94
    640x480       75.00    72.81    59.94    59.93
    640x350       70.10
DP-0 disconnected (normal left inverted right x axis y axis)
DP-1 disconnected (normal left inverted right x axis y axis)
DP-2 disconnected (normal left inverted right x axis y axis)
DP-3 disconnected (normal left inverted right x axis y axis)
```

**Comment by SneWs** [comment]
Even if the system is unable to detect the correct thing here it should not override a user set settings value. That just sounds wrong to me.

**Comment by djordan2** [comment]
Okay, I see why the default behavior is 100% as far as the dpi goes.  You're right that it should respect your settings though.  Could you provide your ~/.config/monitors.xml file?  That contains the all the monitor configurations the system should be remembering.

**Comment by SneWs** [comment]
`
cat ~/.config/monitors.xml
<monitors version="2">
  <configuration>
    <logicalmonitor>
      <x>3840</x>
      <y>0</y>
      <scale>2</scale>
      <monitor>
        <monitorspec>
          <connector>HDMI-0</connector>
          <vendor>AUS</vendor>
          <product>ASUS PB27U</product>
          <serial>0x00009bfa</serial>
        </monitorspec>
        <mode>
          <width>3840</width>
          <height>2160</height>
          <rate>60</rate>
        </mode>
      </monitor>
    </logicalmonitor>
    <logicalmonitor>
      <x>0</x>
      <y>0</y>
      <scale>2</scale>
      <primary>yes</primary>
      <monitor>
        <monitorspec>
          <connector>HDMI-1</connector>
          <vendor>AUS</vendor>
          <product>ASUS PB27U</product>
          <serial>0x00009c02</serial>
        </monitorspec>
        <mode>
          <width>3840</width>
          <height>2160</height>
          <rate>60</rate>
        </mode>
      </monitor>
    </logicalmonitor>
  </configuration>
</monitors>
`

**Comment by djordan2** [comment]
Hi, I've done some testing to look at the cases where this comes up.  It seems to happen with monitors that aren't quite HiDPI according to the cutoffs used by either gnome (192) or the daemon (170), and the daemon is doing pixel-doubling.  

@SneWs I know you already provided the output of xrandr, but I suspect that was after you'd already set scaling in control center.  Could you provide xrandr one more time as it is while everything still looks like it's 100% scale factor?  It think the scale factor may be at 200% but with the monitors set rendered at 8k resolution.

If that's the case this should be a straightforward bug.

**Comment by SneWs** [comment]
You are correct. Below you'll find the output from xrandr after logging in to a fresh session where it's been reset to 100% scaling.

```
xrandr 
Screen 0: minimum 8 x 8, current 7680 x 2160, maximum 32767 x 32767
DVI-D-0 disconnected (normal left inverted right x axis y axis)
HDMI-0 connected 3840x2160+3840+0 (normal left inverted right x axis y axis) 596mm x 335mm
   3840x2160     60.00*+  59.94    50.00    30.00    29.97    25.00    23.98  
   2560x1600     59.97  
   2560x1440     59.95  
   1920x1080     60.00    59.94    50.00    29.97    25.00    23.98    60.00    50.04  
   1680x1050     59.95  
   1440x900      59.89  
   1440x576      50.00  
   1440x480      59.94  
   1280x1024     75.02    60.02  
   1280x960      60.00  
   1280x720      60.00    59.94    50.00  
   1152x864      75.00  
   1024x768      75.03    70.07    60.00  
   800x600       75.00    72.19    60.32    56.25  
   720x576       50.00  
   720x480       59.94  
   640x480       75.00    72.81    59.94    59.93  
   640x350       70.10  
HDMI-1 connected primary 3840x2160+0+0 (normal left inverted right x axis y axis) 596mm x 335mm
   3840x2160     60.00*+  59.94    50.00    30.00    29.97    25.00    23.98  
   2560x1600     59.97  
   2560x1440     59.95  
   1920x1080     60.00    59.94    50.00    29.97    25.00    23.98    60.00    50.04  
   1680x1050     59.95  
   1440x900      59.89  
   1440x576      50.00  
   1440x480      59.94  
   1280x1024     75.02    60.02  
   1280x960      60.00  
   1280x720      60.00    59.94    50.00  
   1152x864      75.00  
   1024x768      75.03    70.07    60.00  
   800x600       75.00    72.19    60.32    56.25  
   720x576       50.00  
   720x480       59.94  
   640x480       75.00    72.81    59.94    59.93  
   640x350       70.10  
DP-0 disconnected (normal left inverted right x axis y axis)
DP-1 disconnected (normal left inverted right x axis y axis)
DP-2 disconnected (normal left inverted right x axis y axis)
DP-3 disconnected (normal left inverted right x axis y axis)
```

**Comment by cdknight** [comment]
Confirming scale settings are not working for monitors that aren't quite hidpi. The solution for me: turn off hidpi-daemon. I filed a similar bug and got similar answers (#226) . Maybe there should be a way to force a monitor as hidpi, not too sure how to implement this though.

**Comment by SneWs** [comment]
Doing some testing. And as a reference point... macOS and Windows properly "detects" the screens as HighDPI screens. But those systems allows for floating point scale values. On Windows for example it automatically sets the screens to a 150% scale factor. And the same goes for macOS.

**Comment by SneWs** [comment]
Good to know, but Wayland is not an option for me currently.

**Comment by djordan2** [comment]
I've identified what's going on here.  The daemon is configuring displays even when there are only non-hipdi monitors connected, which isn't necessary in many (but not all) cases.  I have committed a patch that should help on login, but there is still more work to correctly handle connecting/disconnecting only-lowdpi displays.

**Comment by djordan2** [comment]
@SneWs @cdknight  This issue should be resolved with the latest 18.04.4 release of hidpi-daemon.  Let us know if scaling is remembered on reboot for you.

**Comment by SneWs** [comment]
Out on vacation, when can I expect it to drop in mainline? Already there? If so, will give it a spin once back home. Thx for your work, much appreciated.
